crates/wasi/src/preview2/host/network.rs

```diff
@@ -0,0 +1,185 @@
+use crate::preview2::bindings::sockets::network::{
+    self, ErrorCode, IpAddressFamily, IpSocketAddress, Ipv4Address, Ipv4SocketAddress, Ipv6Address,
+    Ipv6SocketAddress,
+};
+use crate::preview2::network::TableNetworkExt;
+use crate::preview2::{TableError, WasiView};
+use std::io;
+
+impl<T: WasiView> network::Host for T {
+    fn drop_network(&mut self, this: network::Network) -> Result<(), anyhow::Error> {
+        let table = self.table_mut();
+
+        table.delete_network(this)?;
+
+        Ok(())
+    }
+}
+
+impl From<TableError> for network::Error {
+    fn from(error: TableError) -> Self {
+        Self::trap(error.into())
+    }
+}
+
+impl From<io::Error> for network::Error {
+    fn from(error: io::Error) -> Self {
+        match error.kind() {
+            // Errors that we can directly map.
+            io::ErrorKind::PermissionDenied => ErrorCode::AccessDenied,
+            io::ErrorKind::ConnectionRefused => ErrorCode::ConnectionRefused,
+            io::ErrorKind::ConnectionReset => ErrorCode::ConnectionReset,
+            io::ErrorKind::NotConnected => ErrorCode::NotConnected,
+            io::ErrorKind::AddrInUse => ErrorCode::AddressInUse,
+            io::ErrorKind::AddrNotAvailable => ErrorCode::AddressNotBindable,
+            io::ErrorKind::WouldBlock => ErrorCode::WouldBlock,
+            io::ErrorKind::TimedOut => ErrorCode::Timeout,
+            io::ErrorKind::Unsupported => ErrorCode::NotSupported,
+            io::ErrorKind::OutOfMemory => ErrorCode::OutOfMemory,
+
+            // Errors we don't expect to see here.
+            io::ErrorKind::Interrupted | io::ErrorKind::ConnectionAborted => {
+                panic!("transient errors should be skipped")
```

Comment:
Instead of panicking if this conversion fails, can we make this a TryFrom and trap execution instead?